### ISSUE TITLE
Evaluation Scripts

### ISSUE THREAD

**Issue by weizhepei** [willccbb/verifiers#24]
Hi @willccbb, many thanks for open sourcing the great work! Just curious, is there any plan to include the evaluation scripts in the verifiers framework? For example, do you have any pointers to a handy evaluation script for the GSM8K benchmark?  

This would make it easier to validate the effectiveness of the GRPO training pipeline implemented in verifiers and see how much improvement it can achieve compared to the off-the-shelf backbone model.

**Comment by willccbb** [comment]
Want to add something eventually that's more of a permanent solution, yes, but the easiest thing for now is to set an eval dataset in the trainer and set it to run at whatever frequency you want (see any TRL trainer example, works the same here). It'll compute all the same metrics used in training; if one of those is global accuracy, that should be equivalent to running an eval with the Environment harness for tool calls.

Not the most efficient because it doesn't take advantage of multi-GPU vLLM. At some point (hopefully soon) I'll add support for an eval function which can use vLLM on more GPUs and leverage the same rollout logic at higher batch sizes.

**Comment by weizhepei** [comment]
Hi @willccbb, just wondering if you've tried the suggested evaluation setup: 
>the easiest thing for now is to set an eval dataset in the trainer and set it to run at whatever frequency you want 
 
Following your suggestion, I passed the eval dataset to GRPOEnvTrainer and the training logs looked good to me:
![Image](https://github.com/user-attachments/assets/e241435b-5028-4a53-9911-c98078fa9599)

However, when I start a new training with model weights loaded from the above final checkpoint and eval_on_start = True, the eval exact_answer_reward appeared to be inconsistent with the previous training log (i.e., 0.01934 vs. 0.53455):

> {'eval_loss': 0.06218210235238075, 'eval_runtime': 3582.6155, 'eval_samples_per_second': 0.368, 'eval_steps_per_second': 0.002, 'completion_length': 1001.4332042875744, **'rewards/exact_answer_reward_func': 0.019345239037647843**, 'rewards/tool_execution_reward_func': 0.031101194919929617, 'rewards/format_reward_func': 0.06769293253975255, 'rewards/xml_reward_func': 0.045648028541888507, 'reward': 0.16378742633830934, 'reward_std': 0.1650128811597824, 'kl': 0.5488048735119048, 'clip_ratio': 0.0, 'epoch': 0}

Since I was using the same train/eval split, I'd expect eval_on_start in the new run to yield an identical result as the last checkpoint in the previous run. Did I misconfigure anything here? Do you have any thoughts on this discrepancy? Thank you!

**Comment by willccbb** [comment]
hmmm it's hard to say without seeing the code, did you pull or otherwise update the codebase between experiments? are you sure that checkpoint weights are being saved properly? have not spent too much time running downstream evals yet beyond the ones done with eval datasets in the trainer

**Comment by weizhepei** [comment]
Actually I was running the ToolEnv example on GSM8K task with the following script:

```python
# script.py
import verifiers as vf
from verifiers.tools import calculator
from verifiers.prompts import CALCULATOR_FEW_SHOT
from verifiers.trainers import GRPOEnvTrainer
from verifiers.envs import ToolEnv
from verifiers.utils import get_default_grpo_config

model_name = "Qwen/Qwen2.5-3B-Instruct"
model, tokenizer = vf.get_model_and_tokenizer(model_name)

# Initialize tool environment for GSM8K
vf_env = ToolEnv(
    dataset="gsm8k",
    few_shot=CALCULATOR_FEW_SHOT[0],
    tools=[calculator],
    max_steps=5
)


dataset = vf_env.get_dataset()
eval_dataset = vf_env.get_eval_dataset()

rubric = vf_env.get_rubric()

training_args = get_default_grpo_config(
    run_name="gsm8k-calc_" + model_name.split("/")[-1].lower(),
    num_gpus=8
)

trainer = GRPOEnvTrainer(
    model=model,
    processing_class=tokenizer,
    reward_funcs=rubric,
    env=vf_env,
    args=training_args,
    train_dataset=dataset,
    eval_dataset=eval_dataset
)

# print("Dataset size:", len(dataset)) # 7473
# print("Eval dataset size:", len(eval_dataset)) # 1319

trainer.train()
```

The only change I made between experiments is to replace model_name = "Qwen/Qwen2.5-3B-Instruct" with model_name = "LAST_CHECKPOINT_PATH"

Anyways, I appreciate the pointers. Will run some additional exps and double check the results/checkpoints.

**Comment by willccbb** [comment]
see tests/eval_api.py in the latest branch for an alternative approach: you can use "vllm serve ..." to create an OpenAI-compatible client for a training checkpoint and then evaluate that way as well.
